### PR TITLE
Allow libcudfjni to be built as a static library

### PR DESCRIPTION
**Pull request by jlowe**
This adds the ability for the JNI native libraries to be built as static libraries rather than shared libraries by specifying `-DBUILD_SHARED_LIBS=OFF` when configuring with cmake.  This can be useful for external projects that are leveraging the JNI libraries and would like to use an archive that they can link into their shared library.

### CODE REVIEW

File: java/pom.xml [hunk 1]
```diff
@@ -392,6 +392,7 @@
                                     <arg value="-DGPU_ARCHS=${GPU_ARCHS}"/>
                                     <arg value="-DCUDF_JNI_ARROW_STATIC=${CUDF_JNI_ARROW_STATIC}"/>
                                     <arg value="-DCUDF_JNI_LIBCUDF_STATIC=${CUDF_JNI_LIBCUDF_STATIC}"/>
+                                    <arg value="-DBUILD_SHARED_LIBS=ON"/>
```
Comment (mythrocks) [inline java/pom.xml:395]: Should this not use `${BUILD_SHARED_LIBS}`?

Comment (jlowe) [reply java/pom.xml:395]: No, because the Java bindings require a shared library in order to work properly.  Specifying `BUILD_SHARED_LIBS=OFF` here would result in a jar that cannot work.  Specifying `BUILD_SHARED_LIBS=OFF` only makes sense when not trying to build the cudf jar (i.e.: leveraging the code from an external project).

Comment (mythrocks) [reply java/pom.xml:395]: > built as static libraries rather than shared libraries by specifying -DBUILD_SHARED_LIBS=OFF when configuring with cmake.

Ah, so. I should have read the description above more carefully.

File: java/src/main/native/CMakeLists.txt [hunk 2]
```diff
@@ -254,8 +256,15 @@ target_include_directories(
...
-  PROPERTIES BUILD_RPATH "\$ORIGIN" INSTALL_RPATH "\$ORIGIN" # set target compile options
-             CXX_STANDARD 17 CXX_STANDARD_REQUIRED ON CUDA_STANDARD 17 CUDA_STANDARD_REQUIRED ON
+  PROPERTIES BUILD_RPATH "\$ORIGIN"
+             INSTALL_RPATH "\$ORIGIN"
+             # set target compile options
+             CXX_STANDARD 17
+             CXX_STANDARD_REQUIRED ON
+             CUDA_STANDARD 17
+             CUDA_STANDARD_REQUIRED ON
+             POSITION_INDEPENDENT_CODE ON
```
Comment (mythrocks) [inline java/src/main/native/CMakeLists.txt:266]: I think I understand why this change was required, but I'd better confirm.

Before this PR, `-fPIC` was implied already because `libcudfjni.so` was always `SHARED`. But now that we might be building archives, `-fPIC` is no longer implied, as per [CMake docs](https://cmake.org/cmake/help/latest/prop_tgt/POSITION_INDEPENDENT_CODE.html). So this needs to be added explicitly?

Comment (jlowe) [reply java/src/main/native/CMakeLists.txt:266]: Yes, if we don't specify this explicitly then when building a static library instead of shared it will not compile PIC by default.  However we need the code compiled PIC in order to use it in another shared library of an external project.

**Review by mythrocks — APPROVED** [review]
LGTM!

**Comment by jlowe** [comment]
@gpucibot merge
